### PR TITLE
read_fwf with urlopen test GH#26376

### LINKED ISSUE #26376: pd.read_fwf fails with file pointer to url
#### Code Sample, a copy-pastable example if possible

```python
# Your code here

import urllib.request
import pandas as pd

url='ftp://ftp.ncdc.noaa.gov/pub/data/igra/igra2-station-list.txt'
urllib.request.urlretrieve(url,url.split('/')[-1])
igrainv=pd.read_fwf(url.split('/')[-1],widths=(2,1,3,5,9,10,7,4,30,5,5,7),
                        names=('CC','Network','Code','StationId','Latitude','Longitude','Elev','dummy','StationName','From','To','Nrec'))

f=urllib.request.urlopen(url)
igrainv=pd.read_fwf(f,widths=(2,1,3,5,9,10,7,4,30,5,5,7),
                        names=('CC','Network','Code','StationId','Latitude','Longitude','Elev','dummy','StationName','From','To','Nrec'))

```
#### Problem description

The first read_fwf from a file downloaded from the internet works, the second read_fwf of the same file using an url pointer fails.
I get the following error:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 803, in read_fwf
    return _read(filepath_or_buffer, kwds)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 429, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 895, in __init__
    self._make_engine(self.engine)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 1132, in _make_engine
    self._engine = klass(self.f, **self.options)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 3605, in __init__
    PythonParser.__init__(self, f, **kwds)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2238, in __init__
    self.unnamed_cols) = self._infer_columns()
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2614, in _infer_columns
    line = self._buffered_line()
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2689, in _buffered_line
    return self._next_line()
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2794, in _next_line
    orig_line = self._next_iter_line(row_num=self.pos + 1)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 2854, in _next_iter_line
    return next(self.data)
  File "/opt/anaconda3/lib/python3.7/site-packages/pandas/io/parsers.py", line 3589, in __next__
    line = next(self.f)
TypeError: 'addinfourl' object is not an iterator

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-754.10.1.el6.centos.plus.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.4.2
pip: 19.1.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: 0.12.1
IPython: 7.5.0
sphinx: 2.0.1
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: 1.2.1
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: 2.6.2
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.8
lxml.etree: 4.3.3
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.3.3
pymysql: None
psycopg2: 2.7.5 (dt dec pq3 ext lo64)
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by liang3zy22**
- [ ] closes #26376 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

To avoid download file, I just compare the columns.

### CODE REVIEW

**Comment by liang3zy22** [comment]
If I added `check_before_test=True,` in the tm.network decorator then run the test case. Following error will be reported:
```
test_read_fwf.py::test_url_urlopen FAILED
pandas/tests/io/parser/test_read_fwf.py:1015 (test_url_urlopen)
args = (), kwargs = {}

    @wraps(t)
    def wrapper(*args, **kwargs):
        if (
            check_before_test
            and not raise_on_error
>           and not can_connect(url, error_classes)
        ):

../../../_testing/_io.py:222: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../_testing/_io.py:274: in can_connect
    if response.status != 200:
/Users/username/mambaforge/envs/pandas-dev/lib/python3.8/tempfile.py:468: in __getattr__
    a = getattr(file, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <addclosehook at 5811361200 whose fp = <_io.BufferedReader name=-1>>
name = 'status'

    def __getattr__(self, name):
        # Attribute lookups are delegated to the underlying file
        # and cached for non-numeric results
        # (i.e. methods are cached, closed and friends are not)
        file = self.__dict__['file']
>       a = getattr(file, name)
E       AttributeError: '_io.BufferedReader' object has no attribute 'status'

/Users/username/mambaforge/envs/pandas-dev/lib/python3.8/tempfile.py:468: AttributeError
```
It seems like a bug in the tm codes?

File: pandas/tests/io/parser/test_read_fwf.py [hunk 1]
```diff
@@ -1010,3 +1011,49 @@ def test_invalid_dtype_backend():
...
+
+
+@pytest.mark.network
+@tm.network(
+    url=("ftp://ftp.ncdc.noaa.gov/pub/data/igra/igra2-station-list.txt"),
+)
+def test_url_urlopen():
+    url = "ftp://ftp.ncdc.noaa.gov/pub/data/igra/igra2-station-list.txt"
+    with urlopen(url) as f:
+        expected = pd.Index(
```
Comment (mroeschke) [inline pandas/tests/io/parser/test_read_fwf.py:1023]: Can you put this outside the context manager?

File: pandas/tests/io/parser/test_read_fwf.py [hunk 2]
```diff
@@ -1010,3 +1011,49 @@ def test_invalid_dtype_backend():
     )
     with pytest.raises(ValueError, match=msg):
         read_fwf("test", dtype_backend="numpy")
+
+
+@pytest.mark.network
+@tm.network(
+    url=("ftp://ftp.ncdc.noaa.gov/pub/data/igra/igra2-station-list.txt"),
```
Comment (mroeschke) [inline pandas/tests/io/parser/test_read_fwf.py:1018]:
```suggestion
    url="ftp://ftp.ncdc.noaa.gov/pub/data/igra/igra2-station-list.txt",
```

**Comment by liang3zy22** [comment]
I done some diagnosis. url open with ftp will return a [urllib.response.addinfourl](https://docs.python.org/3/library/urllib.request.html#urllib.response.addinfourl) instance. There is a status method of addinfourl, but from python 3.9 only. And I have tried using python 3.10, the status only return None. 

It seems that if the ftpurl can be open, then it is also connectable. So I added codes in can_connect to only check http/https url status.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks for the test and nice investigation @liang3zy22

**Pull request merged**